test/common/formatter/substitution_formatter_test.cc

```diff
@@ -2092,6 +2095,113 @@ TEST(SubstitutionFormatterTest, FilterStateFormatter) {
+  Http::TestResponseHeaderMapImpl response_headers;
+  Http::TestResponseTrailerMapImpl response_trailers;
+  std::string body;
+  auto src_addr =
+      Network::Address::InstanceConstSharedPtr(new Network::Address::Ipv4Instance("1.2.3.4", 773));
+  auto dst_addr =
+      Network::Address::InstanceConstSharedPtr(new Network::Address::Ipv4Instance("0.1.1.2", 513));
+
+  // No parameter value passed in for tlv_type. 
+  {
+    EXPECT_THROW_WITH_MESSAGE(SubstitutionFormatParser::parse("%PROXY_PROTOCOL_TLVS()%"), EnvoyException,
+                              "PROXY_PROTOCOL_TLVS requires parameters");
+  }
+
+  // Invalid tlv_type passed as parameter -- non-integer
+  {
+    EXPECT_THROW_WITH_MESSAGE(SubstitutionFormatParser::parse("%PROXY_PROTOCOL_TLVS(INVALID_TLV_TYPE)%"), EnvoyException,
+                              "Invalid parameter provided for PROXY_PROTOCOL_TLVS header: INVALID_TLV_TYPE. Not parsable as int.");
+  }
+
+  // Invalid tlv_type passed as parameter -- too large of a number for stoi to parse as int
+  {
+    EXPECT_THROW_WITH_MESSAGE(SubstitutionFormatParser::parse("%PROXY_PROTOCOL_TLVS(3000000000)%"), EnvoyException,
+                              "Invalid parameter provided for PROXY_PROTOCOL_TLVS header: 3000000000. Not parsable as int.");
+  }
+
+  // Invalid tlv_type passed as parameter -- too large of a number to be a valid TLV type
+  {
+    EXPECT_THROW_WITH_MESSAGE(SubstitutionFormatParser::parse("%PROXY_PROTOCOL_TLVS(300)%"), EnvoyException,
+                              "Invalid parameter provided for PROXY_PROTOCOL_TLVS header: 300. Not a valid TLV type.");
+  }
+
+  // ProxyProtocolFilterState is not stored in FilterState
+  {
+    NiceMock<StreamInfo::MockStreamInfo> stream_info;
+
+    EXPECT_CALL(Const(stream_info), filterState()).Times(testing::AtLeast(1));
+
+    auto providers = SubstitutionFormatParser::parse("%PROXY_PROTOCOL_TLVS(2)%");
+
+    ASSERT_EQ(providers.size(), 1);
+
+    EXPECT_EQ("", providers[0]->format(request_headers, response_headers, response_trailers,
+                                           stream_info, body));
+  }
+
+  // No TLVs stored in FilterState with specified tlv_type
+  {
+    NiceMock<StreamInfo::MockStreamInfo> stream_info;
+    Network::ProxyProtocolTLV tlv{0x5, "1"};
+    Network::ProxyProtocolData proxy_proto_data{src_addr, dst_addr, {tlv}};
+      stream_info.filter_state_->setData(
+        "envoy.network.proxy_protocol_options",
+          std::make_unique<Network::ProxyProtocolFilterState>(proxy_proto_data),
+          StreamInfo::FilterState::StateType::Mutable, StreamInfo::FilterState::LifeSpan::Connection);
+
+    EXPECT_CALL(Const(stream_info), filterState()).Times(testing::AtLeast(1));
+
+    auto providers = SubstitutionFormatParser::parse("%PROXY_PROTOCOL_TLVS(2)%");
+
+    ASSERT_EQ(providers.size(), 1);
+
+    EXPECT_EQ("", providers[0]->format(request_headers, response_headers, response_trailers,
```

Comment:
Just wondering, in this case, will we not attach the header or attach a header with an empty value?

Reply:
Empty and null values are ignored downstream of the formatter, so no header will be attached.

Reply:
I see. That makes sense. Thanks for the explanation.